File: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Models/GenerationResult.swift
import Foundation

/// Result of a text generation request
public struct GenerationResult {
    /// Generated text (with thinking content removed if extracted)
    public let text: String

    /// Thinking/reasoning content extracted from the response
    public let thinkingContent: String?

    /// Number of tokens used
    public let tokensUsed: Int

    /// Model used for generation
    public let modelUsed: String

    /// Latency in milliseconds
    public let latencyMs: TimeInterval

    /// Execution target (device/cloud/hybrid)
    public let executionTarget: ExecutionTarget

    /// Amount saved by using on-device execution
    public let savedAmount: Double

    /// Framework used for generation (if on-device)
    public let framework: LLMFramework?

    /// Hardware acceleration used
    public let hardwareUsed: HardwareAcceleration

    /// Memory used during generation (in bytes)
    public let memoryUsed: Int64

    /// Tokenizer format used
    public let tokenizerFormat: TokenizerFormat?

    /// Detailed performance metrics
    public let performanceMetrics: PerformanceMetrics

    /// Additional metadata
    public let metadata: ResultMetadata?

    /// Structured output validation result (if structured output was requested)
    public var structuredOutputValidation: StructuredOutputValidation?

    /// Initializer
    internal init(
        text: String,
        thinkingContent: String? = nil,
        tokensUsed: Int,
        modelUsed: String,
        latencyMs: TimeInterval,
        executionTarget: ExecutionTarget,
        savedAmount: Double,
        framework: LLMFramework? = nil,
        hardwareUsed: HardwareAcceleration = .cpu,
        memoryUsed: Int64 = 0,
        tokenizerFormat: TokenizerFormat? = nil,
        performanceMetrics: PerformanceMetrics,
        metadata: ResultMetadata? = nil,
        structuredOutputValidation: StructuredOutputValidation? = nil
    ) {
        self.text = text
        self.thinkingContent = thinkingContent
        self.tokensUsed = tokensUsed
        self.modelUsed = modelUsed
        self.latencyMs = latencyMs
        self.executionTarget = executionTarget
        self.savedAmount = savedAmount
        self.framework = framework
        self.hardwareUsed = hardwareUsed
        self.memoryUsed = memoryUsed
        self.tokenizerFormat = tokenizerFormat
        self.performanceMetrics = performanceMetrics
        self.metadata = metadata
        self.structuredOutputValidation = structuredOutputValidation
    }
}

/// Result metadata for additional strongly-typed information
public struct ResultMetadata {
    public let routingReason: RoutingReasonType
    public let fallbackUsed: Bool
    public let cacheHit: Bool
    public let modelVersion: String?
    public let experimentId: String?
    public let debugInfo: DebugInfo?

    public init(
        routingReason: RoutingReasonType,
        fallbackUsed: Bool = false,
        cacheHit: Bool = false,
        modelVersion: String? = nil,
        experimentId: String? = nil,
        debugInfo: DebugInfo? = nil
    ) {
        self.routingReason = routingReason
        self.fallbackUsed = fallbackUsed
        self.cacheHit = cacheHit
        self.modelVersion = modelVersion
        self.experimentId = experimentId
        self.debugInfo = debugInfo
    }
}

/// Strongly typed routing reason
public enum RoutingReasonType {
    case userPreference
    case costOptimization
    case performanceOptimization
    case resourceConstraint
    case policyDriven
    case fallback
    case experimental
}

/// Debug information for development
public struct DebugInfo {
    public let startTime: Date
    public let endTime: Date
    public let threadCount: Int
    public let deviceLoad: DeviceLoadLevel

    public init(startTime: Date, endTime: Date, threadCount: Int, deviceLoad: DeviceLoadLevel) {
        self.startTime = startTime
        self.endTime = endTime
        self.threadCount = threadCount
        self.deviceLoad = deviceLoad
    }
}

/// Device load level
public enum DeviceLoadLevel {
    case idle       // 0-20%
    case low        // 20-40%
    case moderate   // 40-60%
    case high       // 60-80%
    case critical   // 80-100%

    public init(percentage: Double) {
        switch percentage {
        case 0..<0.2:
            self = .idle
        case 0.2..<0.4:
            self = .low
        case 0.4..<0.6:
            self = .moderate
        case 0.6..<0.8:
            self = .high
        default:
            self = .critical
        }
    }
}

File: sdk/runanywhere-android/src/main/java/com/runanywhere/sdk/models/GenerationResult.kt
package com.runanywhere.sdk.models

/**
 * Result of a text generation request
 */
data class GenerationResult(
    /**
     * Generated text
     */
    val text: String,
    
    /**
     * Number of tokens used
     */
    val tokensUsed: Int,
    
    /**
     * Model used for generation
     */
    val modelUsed: String,
    
    /**
     * Latency in milliseconds
     */
    val latencyMs: Long,
    
    /**
     * Execution target (device/cloud/hybrid)
     */
    val executionTarget: ExecutionTarget,
    
    /**
     * Amount saved by using on-device execution
     */
    val savedAmount: Double,
    
    /**
     * Framework used for generation (if on-device)
     */
    val framework: LLMFramework? = null,
    
    /**
     * Hardware acceleration used
     */
    val hardwareUsed: HardwareAcceleration = HardwareAcceleration.CPU,
    
    /**
     * Memory used during generation (in bytes)
     */
    val memoryUsed: Long = 0,
    
    /**
     * Tokenizer format used
     */
    val tokenizerFormat: TokenizerFormat? = null,
    
    /**
     * Detailed performance metrics
     */
    val performanceMetrics: PerformanceMetrics,
    
    /**
     * Additional metadata
     */
    val metadata: ResultMetadata? = null
)

/**
 * Result metadata for additional strongly-typed information
 */
data class ResultMetadata(
    val routingReason: RoutingReasonType,
    val fallbackUsed: Boolean = false,
    val cacheHit: Boolean = false,
    val modelVersion: String? = null,
    val experimentId: String? = null,
    val debugInfo: DebugInfo? = null
)

/**
 * Strongly typed routing reason
 */
enum class RoutingReasonType {
    USER_PREFERENCE,
    COST_OPTIMIZATION,
    PERFORMANCE_OPTIMIZATION,
    RESOURCE_CONSTRAINT,
    POLICY_DRIVEN,
    FALLBACK,
    EXPERIMENTAL
}

/**
 * Debug information for development
 */
data class DebugInfo(
    val startTime: Long,
    val endTime: Long,
    val threadCount: Int,
    val deviceLoad: DeviceLoadLevel
)

/**
 * Device load level
 */
enum class DeviceLoadLevel {
    IDLE,       // 0-20%
    LOW,        // 20-40%
    MODERATE,   // 40-60%
    HIGH,       // 60-80%
    CRITICAL;   // 80-100%
    
    companion object {
        fun fromPercentage(percentage: Double): DeviceLoadLevel {
            return when {
                percentage < 0.2 -> IDLE
                percentage < 0.4 -> LOW
                percentage < 0.6 -> MODERATE
                percentage < 0.8 -> HIGH
                else -> CRITICAL
            }
        }
    }
}

/**
 * Performance metrics
 */
data class PerformanceMetrics(
    val inferenceTime: Long,
    val tokenizationTime: Long,
    val totalTime: Long,
    val tokensPerSecond: Double,
    val memoryPeak: Long,
    val cpuUsage: Double,
    val gpuUsage: Double? = null
) 